### PR TITLE
Refactor: use channel for FindBestPlanningSolutions function response

### PR DESCRIPTION
**Pull request by timwangmusic**
## Description
The `FindBestPlanningSolutions` method used to return `[]PlanningSolution`. The issue is when time out happens, the iterator's loop has to terminate and no meaningful result will be returned.

## Solution
* Refactor `FindBestPlanningSolutions` method to return `chan PlanningResp`.
* Refactor calling function code.
* Set context with timeout in `FindBestPlanningSolutions` parent function.

## Testing
- [x] Integration testing on Heroku staging
- [x] Updated unit tests

## Checks
- [x] Have you removed commented code?
- [x] Have you used gofmt to format your code?

### CODE REVIEW

File: planner/solver.go [hunk 1]
```diff
@@ -340,20 +337,24 @@ func (s *Solver) generateSolutions(ctx context.Context, req *PlanningReq, timeMa
...
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, SolverTimeout)
+	defer cancel()
+
+	select {
+	case <-ctxWithTimeout.Done():
+		resp.Err = errors.New("cannot complete computation in time")
 		resp.ErrorCode = RequestTimeOut
-		return resp, err
+		return
+	case r := <-FindBestPlanningSolutions(ctxWithTimeout, placeClusters, req.NumPlans, mdIter):
```
Comment (rwangsc18) [inline planner/solver.go:355]: maybe we can also use `resp` here? case resp <- FindBestPlanningSolutions()

File: planner/solver.go [hunk 2]
```diff
@@ -340,20 +337,24 @@ func (s *Solver) generateSolutions(ctx context.Context, req *PlanningReq, timeMa
...
+		return r
```
Comment (rwangsc18) [inline planner/solver.go:356]: then we can simply write `return` here?

File: planner/solver.go [hunk 3]
```diff
@@ -231,18 +224,20 @@ func createPlanningSolutionCandidate(placeIndexes []int, placeClusters [][]match
 	return res, nil
 }
 
-func FindBestPlanningSolutions(ctx context.Context, placeClusters [][]matching.Place, topSolutionsCount int, iterator *MultiDimIterator) ([]PlanningSolution, error) {
+func FindBestPlanningSolutions(ctx context.Context, placeClusters [][]matching.Place, topSolutionsCount int, iterator *MultiDimIterator) (resp chan PlanningResp) {
```
Comment (rwangsc18) [inline planner/solver.go:227]: we can explicitly specify that `resp` is a **send-only** channel in the function definition, so we prevent read operation in the function.

(resp chan<- PlanningResp)

**Review by rwangsc18 — APPROVED** [review]
LGTM

File: planner/solver.go (hunk 3 above)
Comment (timwangmusic) [reply planner/solver.go:227]: if we do that then the caller `generateSolutions` cannot use the channel.

**Pull request merged**